### DISCUSSION TITLE
Inconsistency with environment variables in middleware

### DISCUSSION THREAD

**Discussion by motdotla** [vercel/next.js#39705] [Help]
This seems like an inconsistency in how Vercel handles environment variables for middleware.

1. In local development, ANY environment variable in your local .env file is made available to middleware.ts files
2. In dev/branch/production environment on Vercel, only environment variables in your .env file prefaced with NEXT_PUBLIC_ are made available to middleware.ts files
3. In dev/branch/production environment on Vercel, any Vercel created (from their interface rather than injected by Dotenv) are made available to middleware.ts files

This is was an unpleasant surprise.

Is there a design decision for this or is it just something overlooked? I would expect middleware to have access to all envs in my packaged .env file.

Welcome any and all thoughts.

**Comment by lucasconstantino** [comment]
Please, this is quite relevant.

**Reply by IsiteYves** [reply]
It appears that there is an inconsistency in how Vercel handles environment variables for middleware, which can be frustrating. In local development, all environment variables in the .env file are available to middleware.ts files. However, in the dev/branch/production environment on Vercel, only variables prefixed with NEXT_PUBLIC_ are accessible to middleware.ts files, along with the ones created directly in the Vercel interface.

This behavior may be a deliberate design decision or an oversight. It would be expected for middleware to have access to all environment variables in the packaged .env file. It would be helpful to have clarification from the Vercel team regarding this discrepancy.

Your observations and concerns are valid, and it would be beneficial to gather more thoughts and insights from the community to better understand this behavior and its implications.

Please note that this response is a suggested draft and can be adjusted based on the specific context and audience on GitHub.

**Comment by lucasconstantino** [comment]
Okay, workaround solution found. First things first:

**There is no inconsistency**
When Vercel deploys the bundled code to lambdas or edge functions, it doesn't ship the `.env` files. As most of you may know already, during the build process Next.js uses the Webpack [`DefinePlugin`](https://webpack.js.org/plugins/define-plugin/) to _replace_ environment variable values in the locations of use – thus why you cannot have dynamic reads from `process.env`. But, when executing locally – either by `next dev` or `next start` after a build – the Node server that boots does read the variables from `.env`, thus why it works locally. If you first run `next build` then remove `.env` files and run `next start`, you'll be able to reproduce the behavior of Vercel deploys locally.

Probably for security concerns, Next.js treats edge targeted bundles (middlewares & server components, currently) the same as it treats client bundles: it only include variables prefixed with `NEXT_PUBLIC_`.

**Workaround**
You can take things by your own hands and add another `DefinePlugin` during edge runtime builds to replace any variables you want, and being as smart as needed for that:

```js
// next.config.js

const config = {
  webpack: (config, { nextRuntime }) => {
    if (nextRuntime === 'edge') {
      config.plugins.push(
        new DefinePlugin({
          'process.env.VARIABLE_NAME': `"${process.env.VARIABLE_NAME}"`,
        })
      )
    }
  },
}

module.exports = config
```

**Reply by luistak** [reply]
Thank you for your workaround but do you know how to handle the same issue to runtime envs?

I have a Next app deployed with k8s and the environment variables are in the pod when I access and run `printenv` but on runtime next always return `undefined` to any `env` I tried to create an API route but it also doesn't have access to envs in runtime it seems the `global.process.env`or `process.env` are always empty.

Premisses:
- I'm not using publicRuntime or serverRuntime since I'm using React Server Components
- I tried to get the env vars from the server component and pass via props to a client component (failed)
- I tried to get the env var from a `layout` component and pass via props (failed)
- I tried an API route like this:

```ts
import { NextRequest, NextResponse } from 'next/server';

const ENVS_STATIC = {
  something: global.process.env.SOMETHING || '',
  trackjsApp: global.process.env.NEXT_PUBLIC_TRACKJS_APPLICATION || '',
  NX_trackjsApp: global.process.env.NX_NEXT_PUBLIC_TRACKJS_APPLICATION || '',
};

export async function GET(req: NextRequest) {
  const envs = {
    something: global.process.env.SOMETHING || '',
    trackjsApp: global.process.env.NEXT_PUBLIC_TRACKJS_APPLICATION || '',
  };

  return NextResponse.json({ envs, ENVS_STATIC });
}
```

And it always returns `undefined`

**Reply by jmurty** [reply]
Thanks very much for this work-around, it helped me solve a confusing problem today.

It's frustrating that there is no official documentation (that I could find) about this limitation with which environment variables are or are not made available to middleware when deployed on Vercel.

For now your work-around seems like a much less risky thing to do than adding the `NEXT_PUBLIC_` prefix to a bunch of sensitive envvars, which would also solve the problem but could increase the chances of them being mistakenly exposed directly in visitors' web browsers – i.e. to the public.

But I wonder why this limitation exists: is it a bug, or for a good reason that isn't explained? If the limitation is for security reasons as you suggest, then forcibly passing environment variables through to middleware/edge functions might be unsafe no matter which mechanism we use.

**Comment by motdotla** [comment]
For others coming across this page that are also using .env.vault files you can now do this:

1. Preface your next commands with dotenv preloading.

```json
"scripts": {
  "dev": "node -r dotenv/config ./node_modules/.bin/next dev",
  "build": "node -r dotenv/config ./node_modules/.bin/next build",
  "start": "node -r dotenv/config ./node_modules/.bin/next start",
  "lint": "node -r dotenv/config ./node_modules/.bin/next lint"
},
```

2. Generate your .env.vault file

```
touch .env.production
# place prod envs there prefaced with NEXT_PUBLIC_
```

```
npx dotenv-vault@latest local build
```

3. Preface your build or run command with the `DOTENV_KEY` in .env.keys file.

```
DOTENV_KEY="dotenv://:key_1234@dotenv.local/vault/.env.vault?environment=development" npm run dev
```

Find complete details here: https://www.dotenv.org/blog/2023/05/22/use-dotenv-vault-with-nextjs.html

and read more about the new .env.vault mechanism here: https://github.com/motdotla/dotenv#-deploying

**Comment by philwolstenholme** [comment]
Is this still the case in September 2023? https://vercel.com/changelog/more-flexible-environment-variables-in-edge-functions-and-middleware hasn't changed anything, presumably?

I assumed the build process for middleware would read from `.env.production` like [the rest of Next.js](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#default-environment-variables) but I suppose that only works for Serverless functions or a Node.js server started with `next start`, not for edge functions or edge middleware.

**Comment by motdotla** [comment]
@philwolstenholme this is still the case unfortunately.

We are using this workaround with the new .env.vault file mechanism: https://www.dotenv.org/docs/frameworks/nextjs/vercel

**Comment by Vibeesarma** [comment]
If you copy and paste your env and to vercel env then it will work

**Comment by ryanolson-aumni** [comment]
I'm running into a situation where I _need_ access to a private environment variable. Why is this not working as one would expect?
